### PR TITLE
Add cargo-bdd diagnostic subcommand

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- expose runtime step registry via `dump_registry` and `--dump-steps`
- gather steps from test binaries in `cargo-bdd`
- document diagnostic step dumping
- persist step usage across processes via `.rstest-bdd-usage.json`

## Testing
- `make fmt`
- `make lint`
- `make test`
- `make markdownlint`
- `make nixie`

------
https://chatgpt.com/codex/tasks/task_e_68b0e3c276d8832286374643016165ab

## Summary by Sourcery

Add diagnostic tooling to rstest-bdd by recording step usage, exposing registry dumps with a `--dump-steps` flag, and introducing a new cargo subcommand `cargo bdd` to list all steps, unused steps, and duplicates across the workspace.

New Features:
- Expose dump_registry, unused_steps, and duplicate_steps functions in the rstest-bdd library and add a `--dump-steps` startup flag to output the step registry as JSON
- Persist step usage to `target/.rstest-bdd-usage.json` to track executed steps across test processes
- Introduce a new `cargo-bdd` crate as a Cargo subcommand providing `steps`, `unused`, and `duplicates` commands

Enhancements:
- Automatically mark and record steps when they are looked up or run
- Merge JSON dumps from all test binaries to cover the entire workspace’s registry

Build:
- Add `crates/cargo-bdd` to the workspace and update Cargo.toml dependencies (serde, serde_json, cargo_metadata, eyre, ctor)

Documentation:
- Document the `cargo bdd` diagnostic commands in the user guide, design doc, and roadmap

Tests:
- Add unit tests for registry dumping and usage flag semantics
- Add behavioral tests for unused and duplicate step detection in rstest-bdd
- Add smoke tests for the `cargo-bdd` CLI subcommand

### CODE REVIEW

File: crates/cargo-bdd/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                continue;
+            }
+            bail!("test binary {bin} failed: {err}");
+        }
+        let mut parsed: Vec<Step> = serde_json::from_slice(&out.stdout)
+            .with_context(|| format!("invalid JSON from {bin}"))?;
+        steps.append(&mut parsed);
+    }
+    Ok(steps)
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/src/main.rs:144]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: crates/cargo-bdd/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        .wrap_err("failed to build tests")?;
+    if !output.status.success() {
+        bail!("cargo test failed");
+    }
+
+    let mut bins = Vec::new();
+    for line in String::from_utf8_lossy(&output.stdout).lines() {
+        if let Ok(Message::CompilerArtifact(artifact)) = serde_json::from_str::<Message>(line)
+            && artifact.target.kind.iter().any(|k| k == "test")
+            && let Some(exe) = artifact.executable
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/src/main.rs:117]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

**Comment by leynos** [comment]
@coderabbitai pause

File: crates/rstest-bdd/src/registry.rs [hunk 3]
```diff
@@ -84,6 +89,74 @@ static STEP_MAP: LazyLock<HashMap<StepKey, StepFn>> = LazyLock::new(|| {
...
+        }
+    }
+}
+
+fn read_used_from_file() -> HashSet<(StepKeyword, String)> {
+    let mut used = HashSet::new();
+    if let Ok(file) = fs::File::open(usage_file_path()) {
+        let reader = BufReader::new(file);
+        for line in reader.lines().map_while(Result::ok) {
+            if let Ok(rec) = serde_json::from_str::<StepUsage>(&line) {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/registry.rs]: Has this now been resolved?

File: crates/cargo-bdd/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        .any(|t| t.kind.iter().any(|k| k == "test"));
+    if !has_tests {
+        return Ok(Vec::new());
+    }
+
+    let output = Command::new("cargo")
+        .args(["test", "--no-run", "--message-format=json"])
+        .output()
+        .wrap_err("failed to build tests")?;
+    if !output.status.success() {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/src/main.rs]: Has this now been resolved?

File: crates/rstest-bdd/src/registry.rs [hunk 5]
```diff
@@ -84,6 +89,74 @@ static STEP_MAP: LazyLock<HashMap<StepKey, StepFn>> = LazyLock::new(|| {
     map
 });
 
+static USED_STEPS: LazyLock<Mutex<HashSet<StepKey>>> = LazyLock::new(|| Mutex::new(HashSet::new()));
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/registry.rs:96]: Has this now been resolved?

File: crates/cargo-bdd/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    for group in groups.into_values().filter(|g| g.len() > 1) {
+        for step in &group {
+            print_step(step);
+        }
+        println!("---");
+    }
+    Ok(())
+}
+
+fn collect_steps() -> Result<Vec<Step>> {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/src/main.rs]: Has this now been resolved?

File: crates/cargo-bdd/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,167 @@
+//! Command-line diagnostic tooling for rstest-bdd.
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/src/main.rs]: Has this now been resolved?

File: crates/cargo-bdd/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+                bail!(
+                    "cargo test failed for target {} in package {}",
+                    target.name,
+                    package.name
+                );
+            }
+        }
+    }
+    Ok(bins)
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/src/main.rs:237]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: crates/cargo-bdd/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    if bins.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let mut steps = Vec::new();
+    for bin in bins {
+        let out = Command::new(&bin)
+            .arg("--dump-steps")
+            .output()
+            .with_context(|| format!("failed to run test binary {bin}"))?;
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/src/main.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/cargo-bdd/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        if !out.status.success() {
+            let err = String::from_utf8_lossy(&out.stderr);
+            if err.contains("Unrecognized option: 'dump-steps'") {
+                continue;
+            }
+            bail!("test binary {bin} failed: {err}");
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/src/main.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/cargo-bdd/tests/cli.rs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    let output = Command::cargo_bin("cargo-bdd")
+        .expect("binary exists")
+        .current_dir("..")
+        .arg("steps")
+        .output()
+        .expect("runs");
+    assert!(output.status.success());
+    let stdout = str::from_utf8(&output.stdout).expect("utf8");
+    assert!(stdout.contains("an unused step"));
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/tests/cli.rs:17]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/Cargo.toml [hunk 12]
```diff
@@ -21,6 +21,10 @@ inventory.workspace = true
 regex.workspace = true
 thiserror.workspace = true
 hashbrown = "0.15"
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0"
+tempfile = "3.0"
+ctor.workspace = true
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/Cargo.toml]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/src/lib.rs [hunk 13]
```diff
@@ -28,12 +29,31 @@ mod types;
...
+        )]
+        {
+            match dump_registry() {
+                Ok(json) => println!("{json}"),
+                Err(e) => eprintln!("failed to serialise step registry: {e}"),
+            }
+        }
+        std::process::exit(0);
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/lib.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/src/registry.rs [hunk 14]
```diff
@@ -84,6 +89,74 @@ static STEP_MAP: LazyLock<HashMap<StepKey, StepFn>> = LazyLock::new(|| {
...
+        let exe =
+            std::env::current_exe().unwrap_or_else(|e| panic!("resolve current executable: {e}"));
+        let target = exe
+            .ancestors()
+            .find(|p| p.file_name().is_some_and(|n| n == "target"))
+            .unwrap_or_else(|| panic!("binary must live under target directory"));
+        target.join(".rstest-bdd-usage.json")
+    });
+    PATH.clone()
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/registry.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/src/registry.rs [hunk 15]
```diff
@@ -111,8 +187,80 @@ pub fn find_step(keyword: StepKeyword, text: StepText<'_>) -> Option<StepFn> {
...
+pub fn unused_steps() -> Vec<&'static Step> {
+    let used = combined_used();
+    iter::<Step>
+        .into_iter()
+        .filter(|s| {
+            let key = (s.keyword, s.pattern.as_str().to_string());
+            !used.contains(&key)
+        })
+        .collect()
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/registry.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: docs/rstest-bdd-design.md [hunk 16]
```diff
@@ -981,6 +981,17 @@ functionality is implemented:
...
+  cargo subcommand. It queries the runtime step registry and exposes three
+  commands: `list-steps`, `list-unused`, and `list-duplicates`. Step usage is
+  recorded whenever a step is resolved by the registry and appended to a
+  workspace‑wide `target/.rstest-bdd-usage.json` file so usage survives across
+  test processes. This allows the CLI to report definitions that were never
+  exercised and groups of steps that share a keyword and pattern. Because
+  `inventory` operates per binary, the subcommand compiles each test target and
+  executes it with a private `--dump-steps` flag to stream the registry as
+  JSON. The tool merges these dumps so diagnostics cover the entire workspace.
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/rstest-bdd-design.md]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: docs/users-guide.md [hunk 17]
```diff
@@ -313,6 +313,25 @@ added, patterns and examples may change. Meanwhile, adopting `rstest‑bdd` in
...
+  test run. Usage is tracked whenever a step is resolved and persisted in
+  `target/.rstest-bdd-usage.json` so diagnostics can span processes.
+- `cargo bdd list-duplicates` groups step definitions that share the same
+  keyword and pattern, helping to identify accidental copies.
+
+The subcommand builds each test target in the workspace and runs the resulting
+binary with a private `--dump-steps` flag to collect the registered steps as
+JSON. The merged output powers the commands above, helping to keep the step
+library tidy and discover dead code early in the development cycle.
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/src/registry.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/registry.rs:96]: Has this now been resolved?

File: crates/cargo-bdd/src/main.rs [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+    for bin in bins {
+        let out = Command::new(&bin)
+            .arg("--dump-steps")
+            .output()
+            .with_context(|| format!("failed to run test binary {bin}"))?;
+        if !out.status.success() {
+            let err = String::from_utf8_lossy(&out.stderr);
+            if err.contains("Unrecognized option: 'dump-steps'") {
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/src/main.rs]: @codex has this issue been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/cargo-bdd/src/main.rs [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+/// let step = Step {
+///     keyword: "Given".into(),
+///     pattern: "example".into(),
+///     file: "src/example.rs".into(),
+///     line: 42,
+///     used: false,
+/// };
+/// print_step(&step);
+/// ```
+fn print_step(step: &Step) {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/cargo-bdd/src/main.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/Cargo.toml (hunk 12 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/Cargo.toml]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/tests/diagnostic_duplicates.rs [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+step!(StepKeyword::When, "duplicate", one, &[]);
+step!(StepKeyword::When, "duplicate", two, &[]);
+
+#[test]
+fn finds_duplicates() {
+    let groups = duplicate_steps();
+    assert!(groups.iter().any(|g| g.len() == 2));
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/diagnostic_duplicates.rs:40]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/tests/diagnostic_persist_usage.rs [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+
+step!(StepKeyword::Given, "a persisted step", dummy, &[]);
+
+fn usage_file() -> PathBuf {
+    let exe = std::env::current_exe().unwrap_or_else(|e| panic!("resolve current executable: {e}"));
+    exe.ancestors()
+        .find(|p| p.file_name().is_some_and(|n| n == "target"))
+        .unwrap_or_else(|| panic!("binary must live under target directory"))
+        .join(".rstest-bdd-usage.json")
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/diagnostic_persist_usage.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: docs/users-guide.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/src/registry.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/registry.rs:91]: Has this now been resolved?

File: crates/rstest-bdd/tests/diagnostic_unused.rs [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+#[expect(clippy::unnecessary_wraps, reason = "wrapper must match StepFn signature")]
+fn used_wrapper(
+    ctx: &StepContext<'_>,
+    _text: &str,
+    _docstring: Option<&str>,
+    _table: Option<&[&[&str]]>,
+) -> Result<(), StepError> {
+    let _ = ctx;
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/diagnostic_unused.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/Cargo.toml (hunk 12 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/Cargo.toml]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/tests/diagnostic_duplicates.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/diagnostic_duplicates.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: crates/rstest-bdd/tests/dump_registry.rs [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+)]
+fn wrapper(
+    ctx: &StepContext<'_>,
+    _text: &str,
+    _docstring: Option<&str>,
+    _table: Option<&[&[&str]]>,
+) -> Result<(), StepError> {
+    let _ = ctx;
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/dump_registry.rs]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: docs/roadmap.md [hunk 24]
```diff
@@ -131,14 +131,14 @@ improves the developer experience.
...
 
-  - [ ] Create a helper binary or `cargo` subcommand (`cargo bdd`).
+  - [x] Create a helper binary or `cargo` subcommand (`cargo bdd`).
 
-  - [ ] Implement a `list-steps` command to print the entire registered step
+  - [x] Implement a `list-steps` command to print the entire registered step
     registry.
 
-  - [ ] Implement commands to identify unused or duplicate step definitions.
+  - [x] Implement commands to identify unused or duplicate step definitions.
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/roadmap.md]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.

File: docs/users-guide.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md]: @codex has this issue identified in this comment been resolved? Please provide evidence for your assessment.

If work remains, please provide clear self-contained instructions (including rationale and location) in a triple-tilde code-fenced block.
